src/DIRC_geo.cpp

```diff
@@ -107,47 +107,40 @@ static Ref_t createDetector(Detector& desc, xml_h e, SensitiveDetector sens)
+  double lens_radius = sqrt(lens_width * lens_width / 4. + bar_height * bar_height / 4.);
 
-  Box gfbox("fbox", 0.5 * prism_short_edge, 0.5 * prism_width, 0.5 * lens_thickness);
-  Box gcbox("cbox", 0.5 * prism_short_edge, 0.5 * prism_width + 1 * mm, 0.5 * lens_thickness);
+  double lens_min_thickness = 2.0 * mm;
 
-  Position         tTrans1(0.5 * (prism_short_edge + lens_height), 0, lens_thickness - layer12);
-  Position         tTrans0(-0.5 * (prism_short_edge + lens_height), 0, lens_thickness - layer12);
-  SubtractionSolid tubox("tubox", gfbox, gcbox, tTrans1);
-  SubtractionSolid gubox("gubox", tubox, gcbox, tTrans0);
+  double ztrans1 = -lens_thickness / 2. - sqrt(lens_r1 * lens_r1 - lens_radius * lens_radius) + lens_min_thickness;
+  double ztrans2 = -lens_thickness / 2. - sqrt(lens_r2 * lens_r2 - lens_radius * lens_radius) + lens_min_thickness * 2;
 
-  Tube gcylinder1("Cylinder1", 0, lens_r1, 0.5 * prism_width, 0 * deg, 360 * deg);
-  Tube gcylinder2("Cylinder2", 0, lens_r2, 0.5 * prism_width - 0.5 * mm, 0 * deg, 360 * deg);
-  Tube gcylinder1c("Cylinder1c", 0, lens_r1, 0.5 * prism_width + 0.5 * mm, 0 * deg, 360 * deg);
-  Tube gcylinder2c("Cylinder2c", 0, lens_r2, 0.5 * prism_width + 0.5 * mm, 0 * deg, 360 * deg);
+  Box  lens_symm_box("lens_symm_box", 0.5 * prism_short_edge, 0.5 * lens_width, 0.5 * lens_thickness);
+  Tube lens_symm_tube("lens_symm_tube", 0, lens_radius, 0.5 * lens_thickness);
 
-  IntersectionSolid lens_layer1_solid("lens_layer1_solid", gubox, gcylinder1,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans1)));
-  SubtractionSolid  gLenst("temp", gubox, gcylinder1c, Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans1)));
+  Sphere lens_sphere1("lens_sphere1", 0, lens_r1);
+  Sphere lens_sphere2("lens_sphere2", 0, lens_r2);
 
-  IntersectionSolid lens_layer2_solid("lens_layer2_solid", gLenst, gcylinder2,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans2)));
-  SubtractionSolid  lens_layer3_solid("lens_layer3_solid", gLenst, gcylinder2c,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans2)));
+  IntersectionSolid lens_box("lens_box", lens_symm_box, lens_symm_box, Position(0, 0, -lens_min_thickness * 2));
+  IntersectionSolid lens_tube("sbox", lens_symm_tube, lens_symm_box, Position(0, 0, lens_min_thickness * 2));
+  UnionSolid        lens_box_tube("lens_box_tube", lens_box, lens_tube);
+
+  IntersectionSolid lens_layer1_solid("lens_layer1_solid", lens_box_tube, lens_sphere1, Position(0, 0, -ztrans1));
+  SubtractionSolid  lens_layer23_solid("lens_layer23_solid", lens_box_tube, lens_sphere1, Position(0, 0, -ztrans1));
+
+  IntersectionSolid lens_layer2_solid("lens_layer2_solid", lens_layer23_solid, lens_sphere2, Position(0, 0, -ztrans2));
+  SubtractionSolid  lens_layer3_solid("lens_layer3_solid", lens_layer23_solid, lens_sphere2, Position(0, 0, -ztrans2));
```

Comment:
@niwgit Up to now, no actual changes to the construction to geometry, just clarifications. But now we can see why this is worrisome: any (photon or other) track that is outside the lens needs to evaluate whether it is in the lens during navigation. That requires checking this chain (due to non-trivial shifts in essentially all of them):
- lens_layer1_solid (intersection solid)
  - lens_box_tube (union solid)
    - lens_box (intersection solid)
      - lens_symm_box
      - lens_symm_box
    - lens_tube
      - lens_symm_tube
      - lens_symm_box
  - lens_sphere1
- lens_layer2_solid (intersection solid)
  - lens_layer23_solid (subtraction solid)
    - lens_box_tube (union solid)
      - lens_box (intersection solid)
        - lens_symm_box
        - lens_symm_box
      - lens_tube
        - lens_symm_tube
        - lens_symm_box
    - lens_sphere1
  - lens_sphere2
- lens_layer3_solid (subtraction solid)
  - lens_layer23_solid
    - lens_box_tube (union solid)
      - lens_box (intersection solid)
        - lens_symm_box
        - lens_symm_box
      - lens_tube
        - lens_symm_tube
        - lens_symm_box
    - lens_sphere1
  - lens_sphere2

That's an large number of volumes to check. This must be done for all photons that escape the bar anywhere along its length. Due to the boolean solids, there is not that much that can be done to speed this up (especially due to the spheres that define the lenses, actually).

What I am suggesting is, for starters, to place the whole lens (three layers) inside the envelope volume `lens_box_tube`, which is the union of all three layers. For any tracks outside the lens, the lookup then reverts to:
- lens_box_tube (union solid)
  - lens_box (intersection solid)
    - lens_symm_box
    - lens_symm_box
  - lens_tube
    - lens_symm_tube
    - lens_symm_box

We can go even further (possibly), by not just defining `lens_box_tube` as the envelope volume, but `lens_symm_box`, the symmetric box around the center of the lens. Now the lookup reverts to:
- lens_symm_box

I assume you can see why we would prefer a geometry lookup of 1 volume instead of 31 volumes.

Reply:
@wdconinc thanks for the suggestions. I'll look into using "lens_symm_box" as the envelope for three layers of lens.

Reply:
The current geometry of lens doesn't seem to work with dd_web_display command. It seems phi angle range should be defined for "lens_symm_tube".

dd_web_display --export -o epic_dirc_test.root $DETECTOR_PATH/epic_dirc_only.xml

Compact          ERROR ++ FAILED    to convert subdetector: cb_DIRC: dd4hep: Evaluator::Object : unknown variable : lens_symm_tube : value=lens_symm_tube [Evaluation error]
terminate called after throwing an instance of 'std::runtime_error'
  what():  dd4hep: Evaluator::Object : unknown variable : lens_symm_tube : value=lens_symm_tube [Evaluation error]
Aborted (core dumped)

Reply:
Strange, it should pick up this: https://github.com/AIDASoft/DD4hep/blob/master/DDCore/include/DD4hep/Shapes.h#L668

Reply:
This seems to be a bug in DD4hep, and it only fails for named tubes and spheres. Anonymous tubes and spheres pick up the default angles correctly.

Reply:
Thanks, I don't see the error message anymore.

Regarding the envelope for lens would it be OK to use a Trap envelope for lens + prism and a Box envelope for bar box?
In the end we need some envelope with air with optical properties for Cherenkov photon propagation.

I have tested envelope of union of Trap + Box in another branch and it looks like in this figure. Would separating this into two envelopes help speed up tracking of optical photons entering the lens?

![prism_vside_view](https://user-images.githubusercontent.com/23059941/232859846-4af3f66e-ccbf-40c8-9e8a-9fcc625cfb69.png)

Reply:
I think that should be fine (if it fits; I think the available space was very tight).

Reply:
I added 3 AirOptical envelopes.
- Box for bars + glue + mirror ("Envelope_box_vol")
- Box for lens ("Envelope_lens_vol")
- Trap for prism + mcp ("Envelope_trap_vol")

But currently there's some overlaps for the Trap envelope. There's a small gap seen between prism and lens also. Maybe I'm missing something when placing prism and mcp volumes.

![prism_issue](https://user-images.githubusercontent.com/23059941/233729124-6ce1ff6b-8aec-4496-a333-60ebf5fbe28c.png)

Reply:
The overlap issue is fixed now.